### PR TITLE
[R]Issue 19 datasets cache

### PR DESCRIPTION
**Pull request by y-moufakkr**

### CODE REVIEW

File: packages/x-bundle/package.json [hunk 1]
```diff
@@ -23,6 +23,17 @@
     "package:generate": "ts-node package.ts"
   },
   "dependencies": {
+    "cache-manager": "^3.6.0",
+    "cache-manager-fs-binary": "^1.0.4",
```
Comment (theodorDiaconu) [inline packages/x-bundle/package.json:27]: isn't this poluting too much? Let's just leave redis memory mongodb and it's enough. This is an "X" bundle. And if they want custom store they can install it and use it.

File: packages/x-bundle/src/XBundle.ts [hunk 2]
```diff
@@ -26,7 +28,7 @@ import { RedisMessenger } from "./services/RedisMessenger";
 import { Messenger as LocalMessenger } from "./services/Messenger";
 import SubscriptionGraphQLModule from "./graphql/subscriptions.graphql-module";
 import { RedisListener } from "./listeners/RedisListener";
-
+const cacheManager = require("cache-manager");
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/XBundle.ts:31]: why not use imports?

File: packages/x-bundle/src/executors/cache.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+) {
+  return async function (_, args, ctx, ast) {
+    //todo make userId parametrable
+    const queryKey = JSON.stringify({ input: args, userId: ctx.userId });
+    const cache = new CacheModule(ctx.container);
+    let result = await cache.get(queryKey);
+    if (result && result.found) return result.data;
+
+    for (let i = 0; i < actions.length; i++) {
+      result = await actions[i](_, args, ctx, ast);
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/executors/cache.ts:16]: try using: for (const action of actions) { ... }

File: packages/x-bundle/src/executors/utils/CacheModule.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import { ContainerInstance, Inject, Service } from "@bluelibs/core";
+import { CACHE_CONFIG, CACHE_MANAGER } from "../../constants";
+import { CacheOptions } from "../../defs";
+export class CacheModule {
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/executors/utils/CacheModule.ts:4]: blank line here pls

File: packages/x-bundle/src/XBundle.ts [hunk 5]
```diff
@@ -37,6 +39,22 @@ export class XBundle extends Bundle<IXBundleConfig> {
...
+    cacheConfig: {
+      store: "memory",
+      storeConfig: {
+        memory: 100,
+        ttl: 60,
+      },
+      ttl: 60 /*seconds*/,
+      isCacheableValue: () => true,
+      refreshThreshold: 1,
+      userBoundness: true,
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/XBundle.ts:51]: shouldn't userBoundness be configured at resolver level? I want some queries to be user bounds, others not, or are these default values? If so, pls use "resolverDefaults: {}"

File: packages/x-bundle/src/validators/declarations.ts [hunk 6]
```diff
@@ -21,10 +21,10 @@ declare module "yup" {
      */
     uniqueField(config?: IUniqueFieldValidationConfig): NumberSchema;
   }
-
+  //@ts-ignore
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/validators/declarations.ts:24]: why?

Comment (y-moufakkr) [reply packages/x-bundle/src/validators/declarations.ts:24]: had some type error blocking my unit testing in my loca, just forgot not to delete it XD

File: packages/x-bundle/DOCUMENTATION.md [hunk 7]
```diff
@@ -970,6 +970,116 @@ export default {
...
+        userBoundness: true,
+        userBoundnessFields: ["userId"],
+        expirationBoundness: true,
+        expirationBoundnessField: "expiredAt", //secondsCount or DateTime,
+        refresh: false,
+      },
+    }
+```
+
+the "resolverDefaultConfig" is the default configuration for the caching options, that you can override in specific resolvers by passing other values in options,
```
Comment (theodorDiaconu) [inline packages/x-bundle/DOCUMENTATION.md:1046]: capitalise

File: packages/x-bundle/DOCUMENTATION.md [hunk 8]
```diff
@@ -970,6 +970,116 @@ export default {
...
+}
+```
+
+#### Configuration
+
+the integrated default configuration of the caching is:
+
+```ts
+cacheConfig:{
+      store: "memory",
```
Comment (theodorDiaconu) [inline packages/x-bundle/DOCUMENTATION.md:1029]: proper indentation. please add vscode prettier autoformatting. use it like const cacheConfig = { } so it's valid ts

File: packages/x-bundle/src/cache/defs.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+export type MongoDbStoreConfig = {
+  uri: string;
+  options?: {
+    collection?: string;
+    compression?: boolean;
+    poolSize?: number;
+    autoReconnect?: boolean;
+  };
+};
+export type MongooseStoreConfig = {
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/cache/defs.ts:53]: do we still need these types, let's first offer defaeult for mongodb, memory, redis.

Comment (y-moufakkr) [reply packages/x-bundle/src/cache/defs.ts:53]: if we delete them, we r gonna make the type of storeConfig any, is that ok ?

File: packages/x-bundle/src/defs.ts [hunk 10]
```diff
@@ -1,6 +1,8 @@
 import { Constructor, ContainerInstance } from "@bluelibs/core";
 import { IAstToQueryOptions } from "@bluelibs/nova";
+import { type } from "os";
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/defs.ts:3]: where is this used?

File: packages/x-bundle/src/defs.ts [hunk 11]
```diff
@@ -22,6 +24,8 @@ export interface IXBundleConfig {
     debug?: boolean;
     messengerClass?: Constructor<IMessenger>;
   };
+  //cache config
+  cacheConfig: ICacheManagerConfig;
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/defs.ts:28]: shouldn't this be optional?

File: packages/x-bundle/src/executors/cache.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
+import { CacheOptions } from "../cache/defs";
+import { CACHE_SERVICE } from "../constants";
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/executors/cache.ts:2]: please suffix CACHE_SERVICE with TOKEN

File: packages/x-bundle/src/executors/cache.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+export function Cache<T>(
+  actions: ((_: any, args: any, ctx: any, ast: any) => Promise<any>)[],
+  options?: CacheOptions
+) {
+  return async function (_, args, ctx, ast) {
+    //configure right options between default and specifiq
+    options = configureOptions(ctx, options);
+
+    //generate cache key
+    const cacheKey = generateCacheKey(options, ctx, ast);
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/executors/cache.ts:17]: can you move this function into the CacheService ?

File: packages/x-bundle/src/executors/cache.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    keyBody = addUserBoundnessFieldsToKeyObject(
+      options.userBoundnessFields,
+      keyBody,
+      ctx
+    );
+
+  return Hasher.hash(keyBody);
+};
+
+export const calculateTtlWithExpirationBoundness = (
```
Comment (theodorDiaconu) [inline packages/x-bundle/src/executors/cache.ts:69]: Let's move these functions under the CacheService please

File: packages/x-bundle/DOCUMENTATION.md [hunk 15]
```diff
@@ -970,6 +970,117 @@ export default {
...
+          auth_pass: "XXXXX",
+          db: 0,
+          ttl: 600,
+        },
+        resolverDefaultConfig: {
+          ttl: 30,
+          refresh: false,
+          userBoundness: true,
+          userBoundnessFields: ["userId"],
+          expirationBoundness: true,
```
Comment (theodorDiaconu) [inline packages/x-bundle/DOCUMENTATION.md:1075]: Can we explain what this is and give some examples for it?

File: packages/x-bundle/DOCUMENTATION.md [hunk 16]
```diff
@@ -970,6 +970,117 @@ export default {
...
+          port: 6379,
+          auth_pass: "XXXXX",
+          db: 0,
+          ttl: 600,
+        },
+        resolverDefaultConfig: {
+          ttl: 30,
+          refresh: false,
+          userBoundness: true,
+          userBoundnessFields: ["userId"],
```
Comment (theodorDiaconu) [inline packages/x-bundle/DOCUMENTATION.md:1074]: Why is this an array, what happens if there are 2 strings?

File: packages/x-bundle/DOCUMENTATION.md [hunk 17]
```diff
@@ -970,6 +970,117 @@ export default {
...
+          host: "localhost",
+          port: 6379,
+          auth_pass: "XXXXX",
+          db: 0,
+          ttl: 600,
+        },
+        resolverDefaultConfig: {
+          ttl: 30,
+          refresh: false,
+          userBoundness: true,
```
Comment (theodorDiaconu) [inline packages/x-bundle/DOCUMENTATION.md:1073]: I wonder if we can avoid duplications here. Like, what if we simply do:
```
contextDependency: Array<keyof GraphQLContext>
```
This way we may be able to specify a list. And we can leave a README that if  you want the cache to be bound to the user you ouse contextDependency: ['userId'].

File: packages/x-bundle/package.json [hunk 18]
```diff
@@ -23,6 +23,8 @@
     "package:generate": "ts-node package.ts"
   },
   "dependencies": {
+    "@types/cache-manager": "^3.4.2",
```
Comment (theodorDiaconu) [inline packages/x-bundle/package.json:26]: This should be a devDependency.

File: packages/x-bundle/DOCUMENTATION.md (hunk 15 above)
Comment (y-moufakkr) [reply packages/x-bundle/DOCUMENTATION.md:1075]: its in the docs :
 userBoundness : if true: takes into account other fields in order to personalize cache usage for every user/role.. 
 userBoundnessFields if userBoundness is true: the fields we want to build on the user boundness. those fields expected to be in ctx

File: packages/x-bundle/DOCUMENTATION.md (hunk 16 above)
Comment (y-moufakkr) [reply packages/x-bundle/DOCUMENTATION.md:1074]: can take into account userId, or role for example or other fields

File: packages/x-bundle/DOCUMENTATION.md (hunk 17 above)
Comment (y-moufakkr) [reply packages/x-bundle/DOCUMENTATION.md:1073]: didn't quite understood, the current implementation is its take an array of fields taht we r gonna read from the context, is it the same as Array<keyof GraphQLContext> ?

**Pull request merged**